doc/man/ctlinnd.8
.\" Automatically generated by Pod::Man 4.07 (Pod::Simple 3.32)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "CTLINND 8"
.TH CTLINND 8 "2015-09-12" "INN 2.6.3" "InterNetNews Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
ctlinnd \- Control the main InterNetNews daemon
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBctlinnd\fR [\fB\-hs\fR] [\fB\-t\fR \fItimeout\fR] [\fIcommand\fR [\fIargument\fR ...]]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBctlinnd\fR sends a message to the control channel of \fIinnd\fR\|(8), the main
InterNetNews daemon.
.PP
In the normal mode of behavior, the message is sent to the server, which
then performs the requested action and sends back a reply with a text
message and an exit code for \fBctlinnd\fR.  If the server successfully
performed the command, \fBctlinnd\fR will print the reply on standard output
and exit with a status of zero.  If the server could not perform the
command, it will direct \fBctlinnd\fR to exit with a status of one.  By
default, \fBctlinnd\fR will wait forever for a response from \fBinnd\fR; this
can be changed with the \fB\-t\fR flag.
.PP
The \f(CW\*(C`shutdown\*(C'\fR, \f(CW\*(C`xabort\*(C'\fR, and \f(CW\*(C`xexec\*(C'\fR commands do not generate a reply,
since they cause \fBinnd\fR to exit.  After these commands, \fBctlinnd\fR will
always exit silently with a status of zero.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-h\fR" 4
.IX Item "-h"
Prints a command summary.  If a command is given along with the \fB\-h\fR
flag, only the usage for that command will be given.
.IP "\fB\-s\fR" 4
.IX Item "-s"
If the command was successful, don't print the output from \fBinnd\fR.
.IP "\fB\-t\fR \fItimeout\fR" 4
.IX Item "-t timeout"
Specifies how long to wait for the reply from the server, for commands
other than \f(CW\*(C`shutdown\*(C'\fR, \f(CW\*(C`xabort\*(C'\fR, and \f(CW\*(C`xexec\*(C'\fR.  \fItimeout\fR is the number
of seconds to wait.  A value of zero says to wait forever, and a value
less than zero says not to wait at all but just exit immediately with
status zero.  When waiting for a reply, \fBctlinnd\fR will check every two
minutes to be sure the server is still running, to make it less likely
that \fBctlinnd\fR will just hang.
.Sp
The default is zero, indicating that \fBctlinnd\fR should wait forever.
.SH "COMMANDS"
.IX Header "COMMANDS"
Here is the complete list of supported commands.  Note that nearly all
commands have a fixed number of arguments.  If a parameter may be an empty
string, it is still necessary to pass the empty string to \fBctlinnd\fR as an
argument (specified in the shell as two adjacent quotes, like \f(CW\*(Aq\*(Aq\fR).
.IP "addhist \fImessage-id\fR \fIarrival\fR \fIexpires\fR \fIposted\fR \fItoken\fR" 4
.IX Item "addhist message-id arrival expires posted token"
Add an entry to the history database for \fImessage-id\fR.  The angle
brackets around the message-ID are optional.  It should normally be
protected from the shell with single quotes.
.Sp
\&\fIarrival\fR, \fIexpires\fR, and \fIposted\fR should be the number of seconds
since epoch and indicate when the article arrived, when it expires (via
the Expires: header), and when it was posted (given in the Date: header),
respectively.  \fIexpires\fR should be \f(CW0\fR if the article doesn't have an
Expires: header.  \fItoken\fR is the storage \s-1API\s0 token for the article, and
may be empty.
.Sp
This command can only be used while the server is running, and will be
rejected if the server is paused or throttled.
.IP "allow \fIreason\fR" 4
.IX Item "allow reason"
Allow remote connections, reversing a prior \f(CW\*(C`reject\*(C'\fR command.  \fIreason\fR
must be the same text given to the \f(CW\*(C`reject\*(C'\fR command, or the empty string
(which matches any reason).
.IP "begin \fIsite\fR" 4
.IX Item "begin site"
Begin feeding \fIsite\fR.  The server will rescan the \fInewsfeeds\fR file to
find the specified site and set up a newsfeed for it.  If the site already
existed, a \f(CW\*(C`drop\*(C'\fR for that site is done first.  This command is
forwarded; see \s-1NOTES\s0 below.
.IP "cancel \fImessage-id\fR" 4
.IX Item "cancel message-id"
Remove the article with the specified message-ID from the local system.
This does not generate a cancel control message; it only affects the local
system.  The angle brackets around the message-ID are optional.  It should
normally be protected from the shell with single quotes (and not double
quotes).  For instance:
.Sp
.Vb 1
\&    ctlinnd cancel \*(Aqtest@foo.bar\*(Aq
.Ve
.Sp
Note that the history database is updated with the specified message-ID
so if an article with the same message-ID is afterwards received, it will
be rejected; it is useful for rejecting spam before it arrives.
.Sp
If the server is throttled manually, this command causes it to briefly
open the history database.  If the server is paused or throttled for any
other reason, this command is rejected.
.IP "changegroup \fIgroup\fR \fIstatus\fR" 4
.IX Item "changegroup group status"
The newsgroup \fIgroup\fR is changed so that its status (the fourth field in
the \fIactive\fR file) becomes \fIstatus\fR.  This may be used to make an existing
group moderated or unmoderated, for example.
.Sp
This command, unlike \f(CW\*(C`newgroup\*(C'\fR or \f(CW\*(C`rmgroup\*(C'\fR, can only be used while the
server is running, and will be rejected if the server is paused or
throttled.
.IP "checkfile" 4
.IX Item "checkfile"
Check the syntax of the \fInewsfeeds\fR file and display a message if any
errors are found.  The details of the errors are reported to syslog.
.IP "drop \fIsite\fR" 4
.IX Item "drop site"
Flush and drop \fIsite\fR from the server's list of active feeds.  This
command is forwarded; see \s-1NOTES\s0 below.
.IP "feedinfo \fIsite\fR" 4
.IX Item "feedinfo site"
Print detailed information about the state of the feed to \fIsite\fR, or
brief status about all feeds if \fIsite\fR is the empty string.
.IP "flush \fIsite\fR" 4
.IX Item "flush site"
Flush the buffer for the specified site.  The action taken depends on the
type of feed the site receives; see \fInewsfeeds\fR\|(5) for more information.
This is useful when the site is being fed by a file and batching is about
to start, or to cleanly shut down and respawn a channel feed.  If \fIsite\fR
is an empty string, all sites are flushed and the \fIactive\fR file and
history database are also flushed to disk.  This command is forwarded; see
\&\s-1NOTES\s0 below.
.Sp
Flushing the \fBinnfeed\fR channel feed is the recommended method of
restarting \fBinnfeed\fR to pick up new configuration.  \fBinnd\fR will spawn a
new \fBinnfeed\fR process while the old process shuts down cleanly.
.IP "flushlogs" 4
.IX Item "flushlogs"
Close the news and error log files and rename them to add \f(CW\*(C`.old\*(C'\fR to the
file name, then open fresh news and error logs.  The \fIactive\fR file and
history database are also flushed to disk.  Exploder and process channels
are reopened so that they properly take into account the new log files.
.IP "go \fIreason\fR" 4
.IX Item "go reason"
Re-open the history database and start accepting articles again, reversing
a previous \f(CW\*(C`pause\*(C'\fR or \f(CW\*(C`throttle\*(C'\fR command.  \fIreason\fR must be either the
empty string or match the text that was given to the earlier \f(CW\*(C`pause\*(C'\fR or
\&\f(CW\*(C`throttle\*(C'\fR command.
.Sp
If a \f(CW\*(C`reject\*(C'\fR command is in effect, this will also reverse it by doing
the equivalent of an \f(CW\*(C`allow\*(C'\fR command if the reason matches \fIreason\fR.
Likewise, if a \f(CW\*(C`reserve\*(C'\fR command had been given, this will clear the
reservation if \fIreason\fR matches the text that was given to \f(CW\*(C`reserve\*(C'\fR.
.Sp
The history database is automatically closed on \f(CW\*(C`throttle\*(C'\fR or \f(CW\*(C`pause\*(C'\fR
and reopened on \f(CW\*(C`go\*(C'\fR, so the history database can be replaced during the
pause or throttle without requiring an explicit \f(CW\*(C`reload\*(C'\fR command.  If any
other configuration files or the \fIactive\fR file were modified, a \f(CW\*(C`reload\*(C'\fR
command should be given to force the server to re-read those files.
.Sp
If the server throttled itself because it accumulated too many I/O errors,
this command will reset the error count.
.Sp
If \fBinnd\fR was not started with the \fB\-n y\fR flag, this command also does
the equivalent of a \f(CW\*(C`readers\*(C'\fR command with \f(CW\*(C`yes\*(C'\fR as the flag and
\&\fIreason\fR as the text.
.IP "hangup \fIchannel\fR" 4
.IX Item "hangup channel"
Close the socket for the specified incoming channel.  This may be useful
when an incoming connection appears to be hung (although \fBinnd\fR will
close idle connections automatically after a timeout).
.IP "help [\fIcommand\fR]" 4
.IX Item "help [command]"
Print a command summary for all commands, or just \fIcommand\fR if one is
specified.  This is equivalent to the \fB\-h\fR option.
.IP "kill \fIsignal\fR \fIsite\fR" 4
.IX Item "kill signal site"
Signal \fIsignal\fR is sent to the process underlying the specified site,
which must be a channel or exploder feed.  \fIsignal\fR may be a numeric
signal number or one of \f(CW\*(C`hup\*(C'\fR, \f(CW\*(C`int\*(C'\fR, or \f(CW\*(C`term\*(C'\fR; case is not
significant.
.IP "logmode" 4
.IX Item "logmode"
Cause the server to log its current operating mode to syslog.
.IP "lowmark \fIfile\fR" 4
.IX Item "lowmark file"
Reset the low water marks in the \fIactive\fR file based on the contents of
\&\fIfile\fR.  Each line in \fIfile\fR must be of the form:
.Sp
.Vb 1
\&    group low\-value
.Ve
.Sp
For example:
.Sp
.Vb 1
\&    comp.lang.c++ 243
.Ve
.Sp
This command is mostly used by \fBnews.daily\fR to update the \fIactive\fR file
after nightly expiration.
.IP "mode" 4
.IX Item "mode"
Print the server's operating mode as a multi-line summary of the
parameters and the operating state.  The parameters in the output
correspond to command-line flags to \fBinnd\fR and give the current settings
of those parameters that can be overridden by command-line flags.
.IP "name \fIchannel\fR" 4
.IX Item "name channel"
Print the name and relevant information for the given incoming or outgoing
channel, or for all channels if \fIchannel\fR is an empty string.  The
response is formatted as:
.Sp
.Vb 1
\&    <name>:<number>:<type>:<idle>:<status>
.Ve
.Sp
where <name> is the name of the channel, <number> is its number (generally
the same as the file descriptor assigned to it), <idle> is the idle time
for an \s-1NNTP\s0 channel or the process \s-1ID\s0 for a process channel, and <status>
is the status for \s-1NNTP\s0 channels.
.Sp
The type is one of the following values:
.Sp
.Vb 6
\&    control     Control channel for ctlinnd
\&    file        An outgoing file feed
\&    localconn   Local channel used by nnrpd and rnews for posting
\&    nntp        NNTP channel for remote connections
\&    proc        The process for a process feed
\&    remconn     The channel that accepts new remote connections
.Ve
.Sp
Channel status indicates whether the channel is paused or not.  Nothing is
shown unless the channel is paused, in which case \f(CW\*(C`paused\*(C'\fR is shown.  A
channel will be paused automatically if the number of remote connections
for that label in \fIincoming.conf\fR is greater than \fImax-connections\fR
within \fIhold-time\fR seconds.
.IP "newgroup \fIgroup\fR [\fIstatus\fR [\fIcreator\fR]]" 4
.IX Item "newgroup group [status [creator]]"
Create the specified newsgroup.  The \fIstatus\fR parameter is the fourth field
of the \fIactive\fR file entry, as described in \fIactive\fR\|(5).  If it is not an
equal sign, only the first character is used.  \fIcreator\fR should be the
identity of the person creating the group.  If the newsgroup already
exists, this is equivalent to the \f(CW\*(C`changegroup\*(C'\fR command.
.Sp
\&\fIcreator\fR, encoded in \s-1UTF\-8\s0 if given, may be omitted; if so, it will
default to the newsmaster (as specified at configure time, normally
\&\f(CW\*(C`usenet\*(C'\fR).  \fIstatus\fR may also be omitted; if so, it will default to \f(CW\*(C`y\*(C'\fR
(a normal, unmoderated group).  The combination of defaults make it possible
to use the text of the Control: header directly (although don't do this
without checking the syntactic validity of the header first).
.Sp
This command can only be done while the server is running or throttled
manually.  It will update its internal state when a \f(CW\*(C`go\*(C'\fR command is

NEWS
Changes in 2.6.3

    * Fixed the selection of the elliptic curve to use with OpenSSL 1.1.0 or
      later; NIST P-256 was enforced instead of using the most secure curve.

    * A new inn.conf parameter has been added to fine-tune the cipher suites
      to use with TLS 1.3:  the *tlsciphers13* now permits configuring them.
      A separate cipher suite configuration parameter is needed for TLS 1.3
      because TLS 1.3 cipher suites are not compatible with TLS 1.2, and
      vice-versa.  In order to avoid issues where legacy TLS 1.2 cipher
      suite configuration configured in the *tlsciphers* parameter would
      inadvertently disable all TLS 1.3 cipher suites, the inn.conf
      configuration has been separated out.

    * Fixed a regression since INN 2.6.1 that prevented articles with
      internationalized header fields (that is to say encoded in UTF-8) from
      being posted.

    * Support for Python 3 has been added to INN.  Embedded Python filtering
      and authentication hooks for innd and nnrpd can now use version 3.3.0
      or later of the Python interpreter.  In the 2.x series, version 2.3.0
      or later is still supported.

      When configuring INN with the --with-python flag, the "PYTHON"
      environment variable, when set, is used to select the interpreter to
      embed.  Otherwise, it is searched in standard paths.

      In case you change the Python interpreter to embed, make sure that the
      Python scripts you use are written in the expected syntax for that
      version of the Python interpreter.  Notably, buffer objects have been
      replaced with memoryview objects in Python 3, and UTF-8 encoding now
      really matters for string literals (Python 3 uses bytes and Unicode
      objects).

      INN documentation and samples of Python hooks have been updated to
      provide more examples.

    * When a Python or Perl filter hook rejects an article, innd now
      mentions the reason in response to CHECK and TAKETHIS commands. 
      Previously, the reason was given only for the IHAVE command.

    * nnrpd now properly logs the hostname of clients whose connection
      failed owing to an issue during the negotiation of a TLS session or
      high load average.

Changes in 2.6.2

    * A new *syntaxchecks* parameter has been added in inn.conf.  It permits
      controlling the level of checks performed by innd and nnrpd.  Up to
      now, only one check can be enabled/disabled:  when *laxmid* is
      mentioned in the values of this new parameter, INN accepts Message-IDs
      that contain ".." in the left part, as well as Message-IDs with two
      "@" (such Message-IDs would otherwise be considered as syntactically
      invalid).  See the inn.conf(5) man page for more details.

      The check is disabled by default (*no-laxmid*), which corresponds to
      the legacy behaviour of INN 2.6.1 and earlier.

    * Use of the ovdb_server helper server is now the default when using the
      ovdb overview method, that is to say the default value for the
      *readserver* parameter in ovdb.conf is now set to true.  It improves
      stability and avoids deadlocks, timing issues and corrupted ovdb
      databases.

    * mailpost now removes empty header fields before attempting to post
      articles, and keeps trace of them in the X-Mailpost-Empty-Hdrs: newly
      generated header field body.  Also, mailpost now sanitizes header
      fields with regards to empty continuation header lines.  Thanks to
      Kamil Jonca for these bug reports.

    * A new -z parameter has been added to mailpost to mention a list of
      header fields to remove from the gated message.  Thanks to Dieter
      Stussy for the patch.

    * Fixed a bug in inews that was rejecting articles containing header
      fields whose length exceeded 998 bytes.  This limitation is for the
      length of a single line of a header field (and not for the length of
      the whole header field, as it was wrongly the case).

    * Added support for GnuPG's gpg binary (in addition to gpgv) in
      pgpverify.  Indeed, gpg still validates signatures made with weak
      digest algorithms like MD5 whereas gpgv no longer do.  Thanks to
      Thomas Hochstein for the patch, which permits validating control
      articles for hierarchies that are still using old PGP keys.

    * Added similar support for GnuPG's gpg binary in perl-nocem to validate
      NoCeM notices from issuers who are still using old PGP keys.

    * A few commands listed in the "Control commands to INND" section in
      daily Usenet reports were appearing as a mere letter; all of them are
      now properly converted to meaningful words.

    * The *tlsprotocols* parameter in inn.conf now recognizes the "TLSv1.3"
      value (for OpenSSL versions implementing TLS 1.3, that is to say
      starting from OpenSSL 1.1.1).

    * The buffindexed overview method will now hopefully work properly on
      systems with a native page size larger than 16KB.

    * Other minor bug fixes and documentation improvements.

Changes in 2.6.1

    * nnrpd now uses -0000 as the time zone for Date: and Injection-Date:
      header fields it generates.  It was previously using +0000, wrongly
      systematically indicating a local time zone at Universal Time when
      *localtime* is set to false (which is the default) in readers.conf. 
      The +0000 time zone will now be used only if *localtime* is set to
      true and UTC is really the local time zone of the server.

    * Julien Elie has implemented in nnrpd the new COMPRESS command
      described in draft-murchison-nntp-compress that extends the NNTP
      protocol to allow a connection to be effectively and efficiently
      compressed.  News clients that also support that extension will be
      able to benefit from that bandwidth optimization and improvement in
      speed.  Moreover, using COMPRESS is more secure than TLS-level

frontends/ctlinnd.c
/*  $Id: ctlinnd.c 10153 2017-06-05 12:28:01Z iulius $
**
**  Send control messages to the InterNetNews daemon.
*/

#include "config.h"
#include "clibrary.h"
#include <ctype.h>
#include <errno.h>
#include <sys/stat.h>

#include "inn/innconf.h"
#include "inn/messages.h"
#include "inn/inndcomm.h"
#include "inn/libinn.h"
#include "inn/paths.h"


/*
**  Datatype for an entry in the command table.
*/
typedef struct _COMMAND {
    const char *Command;
    const char *Text;
    int		argc;
    char	Letter;
    bool	Glue;
} COMMAND;


static COMMAND	Commands[] = {
    {	"addhist",	"id arr exp post token...\tAdd history line",
	5,	SC_ADDHIST,	true	},
    {	"allow",	"reason...\t\t\tAllow remote connections",
	1,	SC_ALLOW,	true	},
    {	"begin",	"site\t\t\tStart newly-added site",
	1,	SC_BEGIN,	false	},
    {	"cancel",	"id\t\t\tCancel message locally",
	1,	SC_CANCEL,	false	},
    {	"changegroup",	"group rest\tChange mode of group",
	2,	SC_CHANGEGROUP,	false	},
    {	"checkfile",	"\t\t\tCheck syntax of newsfeeds file",
	0,	SC_CHECKFILE,	false	},
    {	"drop",		"site\t\t\tStop feeding site",
	1,	SC_DROP,	false		},
    {	"feedinfo",		"site\t\t\tPrint state of feed to site*",
	1,	SC_FEEDINFO,	false		},
    {	"flush",	"site\t\t\tFlush feed for site*",
	1,	SC_FLUSH,	false	},
    {	"flushlogs",	"\t\t\tFlush log files",
	0,	SC_FLUSHLOGS,	false	},
    {	"go",		"reason...\t\t\tRestart after pause or throttle",
	1,	SC_GO,		true	},
    {	"hangup",	"channel\t\tHangup specified incoming channel",
	1,	SC_HANGUP,	false	},
    {	"logmode",		"\t\t\t\tSend server mode to syslog",
	0,	SC_LOGMODE,	false		},
    {	"mode",		"\t\t\t\tPrint operating mode",
	0,	SC_MODE,	false		},
    {	"name",		"nnn\t\t\tPrint name of specified channel*",
	1,	SC_NAME,	false		},
    {	"newgroup",	"group rest creator\tCreate new group",
	3,	SC_NEWGROUP,	false	},
    {	"param",	"letter value\t\tChange command-line parameters",
	2,	SC_PARAM,	false	},
    {	"pause",	"reason...\t\tShort-term pause in accepting articles",
	1,	SC_PAUSE,	true	},
#ifdef DO_PERL
    {	"perl",			"flag\t\t\tEnable or disable Perl filtering",
	1,	SC_PERL,	false	},
#endif
#ifdef DO_PYTHON
    {	"python",		"flag\t\t\tEnable or disable Python filtering",
	1,	SC_PYTHON,	false	},
#endif
    {	"readers",	"flag text...\t\tEnable or disable newsreading",
	2,	SC_READERS,	true	},
    {	"reject",	"reason...\t\t\tReject remote connections",
	1,	SC_REJECT,	true	},
    {	"reload",	"what reason...\t\tRe-read config files*",
	2,	SC_RELOAD,	true	},
    {	"renumber",	"group\t\tRenumber the active file*",
	1,	SC_RENUMBER,	false	},
    {	"reserve",	"reason...\t\tReserve the next pause or throttle",
	1,	SC_RESERVE,	true	},
    {	"rmgroup",	"group\t\t\tRemove named group",
	1,	SC_RMGROUP,	false	},
    {	"send",		"feed text...\t\tSend text to exploder feed",
	2,	SC_SEND,	true	},
    {	"shutdown",	"reason...\t\tShut down server",
	1,	SC_SHUTDOWN,	true	},
    {	"stathist",	"filename|off\t\tLog into filename some history stats",
	1,	SC_STATHIST,	false	},
    {	"status",	"interval|off\t\tTurn innd status generation on or off",
	1,	SC_STATUS,	false	},
    {	"kill",	"signal site\t\tSend signal to site's process",
	2,	SC_SIGNAL,	false	},
    {	"throttle",	"reason...\t\tStop accepting articles",
	1,	SC_THROTTLE,	true	},
    {   "timer",	"interval|off\t\tTurn performance monitoring on or off",
	1,	SC_TIMER,	false	},
    {	"trace",	"innd|#|nnrpd flag\tTurn tracing on or off",
	2,	SC_TRACE,	false	},
    {	"xabort",	"text...\t\tAbort the server",
	1,	SC_XABORT,	true	},
    { "lowmark",	"filename\t\tReset active file low article marks",
	1,	SC_LOWMARK,	false	},
    { "renumberlow",	"filename\t\tReset active file low article marks",
	1,	SC_LOWMARK,	false	},
    {	"xexec",	"path\t\t\tExec new server",
	1,	SC_XEXEC,	false	}
};


static void Usage(const char *) __attribute__ ((__noreturn__));
static void WrongArgs(COMMAND *) __attribute__ ((__noreturn__));


/*
**  Print a help summary.
*/
static void
Help(char *p)
{
    COMMAND	*cp;

    if (p == NULL) {
	printf("Command summary:\n");
	for (cp = Commands; cp < ARRAY_END(Commands); cp++)
	    printf("  %s %s\n", cp->Command, cp->Text);
	printf("*   Empty string means all sites/groups/etc.\n");
	printf("... All trailing words are glued together.\n");
	exit(0);
    }
    for (cp = Commands; cp < ARRAY_END(Commands); cp++)
	if (strcmp(p, cp->Command) == 0) {
	    printf("Command usage:\n");
	    printf("  %s %s\n", cp->Command, cp->Text);
	    exit(0);
	}
    printf("No such command.\n");
    exit(0);
}


/*
**  Print a command-usage message and exit.
*/
static void
WrongArgs(COMMAND *cp)
{
    printf("Wrong number of arguments -- usage:\n");
    printf("  %s %s\n", cp->Command, cp->Text);
    exit(1);
}


/*
**  Print an error message and exit.
*/
static void
Failed(const char *p)
{
    if (ICCfailure)
        syswarn("cannot %s (%s failure)", p, ICCfailure);
    else
        syswarn("cannot %s", p);
    ICCclose();
    exit(1);
}


/*
**  Print an error reporting incorrect usage.
*/
static void
Usage(const char *what)
{
    fprintf(stderr, "Usage error (%s) -- try -h for help.\n", what);
    exit(1);
}


int main(int ac, char *av[])
{
    static char		Y[] = NF_FLAG_OK_STRING;
    static char		EMPTY[] = "";
    COMMAND	        *cp;
    char	        *p;
    int	                i;
    bool		Silent;
    bool		NeedHelp;
    char		*reply;
    char		*new;
    int			length;
    char		*nv[4];
    struct stat		Sb;
    char		buff[SMBUF];

    /* First thing, set up our identity. */
    message_program_name = "ctlinnd";

    /* Set defaults. */
    if (!innconf_read(NULL))
        exit(1);
    Silent = false;
    NeedHelp = false;
    ICCsettimeout(CTLINND_TIMEOUT);

    /* Parse JCL. */
    while ((i = getopt(ac, av, "hst:")) != EOF)
	switch (i) {
	default:
	    Usage("bad flags");
	    /* NOTREACHED */
	case 'h':		/* Get help			*/
	    NeedHelp = true;
	    break;
	case 's':		/* Silent -- no output		*/
	    Silent = true;
	    break;
	case 't':		/* Time to wait for reply	*/
	    ICCsettimeout(atoi(optarg));
	    break;
	}
    ac -= optind;
    av += optind;
    if (NeedHelp)
	Help(av[0]);
    if (ac == 0)
	Usage("missing command");

    /* Look up the command word and move to the arguments. */
    if (strcmp(av[0], "help") == 0)
	Help(av[1]);
    for (cp = Commands; cp < ARRAY_END(Commands); cp++)
	if (strcmp(av[0], cp->Command) == 0)
	    break;
    if (cp == ARRAY_END(Commands))
	Usage("unknown command");
    ac--;
    av++;

    /* Check argument count. */
    if (cp->Letter == SC_NEWGROUP) {
	/* Newgroup command has defaults. */
	switch (ac) {
	default:
	    WrongArgs(cp);
	    /* NOTREACHED */
	case 1:
	    nv[0] = av[0];
	    nv[1] = Y;
	    nv[2] = EMPTY;
	    nv[3] = NULL;
	    av = nv;
	    break;
	case 2:
	    nv[0] = av[0];
	    nv[1] = av[1];
	    nv[2] = EMPTY;
	    nv[3] = NULL;
	    av = nv;
	    break;
	case 3:
	    break;
	}
    }
    else if (ac > cp->argc && cp->Glue) {
	/* Glue any extra words together. */
	for (length = 0, i = cp->argc - 1; (p = av[i++]) != NULL; )
	    length += strlen(p) + 1;
        new = xmalloc(length);
        *new = '\0';
	for (i = cp->argc - 1; av[i]; i++) {
	    if (i >= cp->argc)
                strlcat(new, " ", length);
            strlcat(new, av[i], length);
	}
	av[cp->argc - 1] = new;
	av[cp->argc] = NULL;
    }
    else if (ac != cp->argc)
	/* All other commands must have the right number of arguments. */
	WrongArgs(cp);

    /* For newgroup and changegroup, make sure the mode is valid. */
    if (cp->Letter == SC_NEWGROUP || cp->Letter == SC_CHANGEGROUP) {
	switch (av[1][0]) {
	default:
	    Usage("Bad group mode");
	    /* NOTREACHED */
	case NF_FLAG_ALIAS:
	case NF_FLAG_JUNK:
	case NF_FLAG_MODERATED:
	case NF_FLAG_OK:
	case NF_FLAG_NOLOCAL:
	case NF_FLAG_IGNORE:
	    break;
	}
    }

    /* Make sure there are no separators in the parameters. */
    for (i = 0; (p = av[i++]) != NULL; )
	if (strchr(p, SC_SEP) != NULL)
            die("illegal character \\%03o in %s", SC_SEP, p);

    /* Do the real work. */
    if (ICCopen() < 0)
	Failed("setup communication");
    i = ICCcommand(cp->Letter, (const char **) av, &reply);
    if (i < 0) {
	i = errno;
	p = concatpath(innconf->pathrun, INN_PATH_SERVERPID);
	if (stat(p, &Sb) < 0)
            warn("no innd.pid file; did server die?");
        free(p);
	snprintf(buff, sizeof(buff), "send \"%s\" command", cp->Command);
	errno = i;
	Failed(buff);
    }

    if (reply) {
	/* Skip "<exitcode><space>" part of reply. */
	for (p = reply; *p && isdigit((unsigned char) *p); p++)
	    continue;
	while (*p && ISWHITE(*p))
	    p++;
	if (i != 0)
            warn("%s", p);
	else if (!Silent)
	    printf("%s\n", p);
    }

    if (ICCclose() < 0)
	Failed("end communication");

    exit(i);
    /* NOTREACHED */
}
